pandas/core/internals/concat.py

```diff
@@ -319,6 +319,15 @@ def _concatenate_join_units(join_units, concat_axis, copy):
                     concat_values = concat_values.copy()
             else:
                 concat_values = concat_values.copy()
+    elif any(isinstance(t, ExtensionArray) for t in to_concat):
+        # concatting with at least one EA means we are concatting a single column
```

Comment:
I find it odd that the logic is now split here and concat_compat. I would strongly encourage to put this logic there.

Reply:
Having this here actually ensures a clear separation of concerns between `dtypes/concat.py` and `internals/concat.py`. 

The `concat_compat` from `dtypes/concat.py` only needs to worry about concatting 1D arrays (for the case of ExtensionArrays, to be clear), and the code here in internals deals with ensuring the result is the proper dimension depending on how it is put in the BlockManager.

This keeps this BlockManager-related logic inside the internals, and ensures that `concat_compat` / `dtypes/concat.py` doesn't need to care about internals-specific details.

Reply:
my point is you should have a concat_ea which you *call* here (right near concat_compat), otherwise the logic is scatttered. the *dispatching) logic is fine, the problem is the operational logic does not belong here. (alternative is to pass anotherr arg to concat_compat to do what you are doing here).

Reply:
If there is a different `concat_ea`, then at each location where `concat_compat` is called, we would need to do a `if any_ea: concat_ea(..); else: concat_compat(...)`

What do you mean exactly with the "operational logic does not belong here". What part of the code below is the "operational" logic?

Reply:
every bit you added is operational logic.

pls do something like this
```
elif any(isinstance(t, ExtensionArray) for t in to_concat):
    values = concat_ea(...)
```

Reply:
`concat_compat` is called in 13 places in the code base. Do you want to add such `elif` case in all those places?

Reply:
this is moving L324-330. As I have said before you are mixing 2 types of logic here by adding this.

Reply:
To make sure I understand you correctly. You want to see the following function be put in `dtypes/concat.py`:

```python
def concat_ea_frame(to_concat):
    to_concat = [t if isinstance(t, ExtensionArray) else t[0, :] for t in to_concat]
    concat_values = concat_compat(to_concat, axis=concat_axis)
    if not isinstance(concat_values, ExtensionArray):
        # if the result of concat is not an EA but an ndarray, reshape to
        # 2D to put it a non-EA Block
        concat_values = np.atleast_2d(concat_values)
    return concat_values
```

and called her. Is that correct?

I can do that if it ends this discussion, but just to note: this would create a function in `dtypes/concat.py` with logic tied to the internals (reshaping to conform to block's dimenions is something only needs to be known to the internals code) and that would only be used in the internals (here, in `internals/concat.py`)

Reply:
yes this is exactly inline with concat_dateime and concat_compat. I am surprised that you think this should be anywhere else.

Reply:
> with logic tied to the internals (reshaping to conform to block's dimenions is something only needs to be known to the internals code) and that would only be used in the internals

i agree with Joris; internals-specific logic belongs in internals

Reply:
i guess